### ISSUE TITLE
Looking for read-only policy for authenticated users

### ISSUE THREAD

**Issue by elvogel** [NuGet/NuGetGallery#5980]
What is the most efficient way to establish a policy where **authenticated users** can download packages directly or through the API, and only administrators can load packages into the platform? Is that kind of policy built into the system, and if not, are there any suggestions for implementation?

Thanks!

**Comment by ryuyu** [comment]
This scenario is not supported out of the box.
For admin-only upload, you can add a check for if currentUser.IsAdministrator to the Upload routes (https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Controllers/PackagesController.cs#L157 and https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Controllers/ApiController.cs#L342)

Adding authorization to Api Download endpoint should be relatively simple, adding an [ApiAuthorize] attribute instead. https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Controllers/ApiController.cs#L136 However, this endpoint is also used in the UI. Since the UI expects a different kind of authorization, this will likely break this scenario. For UI, you will likely have to implement a new route that has a [UIAuthorize] attribute and is functionally identical to the Api route.

Do note that these solutions may not work if you want to use these as feeds in VS or nuget.exe.

Can you give us a little more information on why you want to do this? What is the end to end scenario you are trying to achieve with these changes?

**Comment by elvogel** [comment]
I've got some commercial libraries that I want to be able to distribute as NuGet packages from a publicly available NuGet portal, but I need to be able to keep track of who's downloading the packages. I'd love to give users the freedom to use VS to download packages, so long as I know who is doing it (and, ideally, keep track of their usage too). They're also .NET Core libraries, so finding a decent licensing tool for them is hard to find. I'd rather just keep track of who's downloading packages.

I've managed to block package upload by changing https://github.com/endpointsystems/NuGetGallery/blob/master/src/NuGetGallery/Services/ActionsRequiringPermissions.cs to prevent anyone that isn't an administrator from uploading packages, so that's a good start.
